paymentsheet/src/main/java/com/stripe/android/customersheet/CustomerAdapterResult.kt

```diff
@@ -0,0 +1,85 @@
+package com.stripe.android.customersheet
+
+import androidx.annotation.RestrictTo
+
+@ExperimentalCustomerSheetApi
+@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
+sealed interface CustomerAdapterResult<T> {
+
+    @ExperimentalCustomerSheetApi
+    @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
+    class Success<T>(val value: T) : CustomerAdapterResult<T>
+
+    @ExperimentalCustomerSheetApi
+    @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
+    class Failure<T>(
+        val cause: Throwable?,
+        val displayMessage: String? = null
+    ) : CustomerAdapterResult<T>
+}
+
+@OptIn(ExperimentalCustomerSheetApi::class)
+internal fun <T> CustomerAdapterResult<T>.getOrNull(): T? {
+    return toResult().getOrNull()
+}
+
+@OptIn(ExperimentalCustomerSheetApi::class)
+internal inline fun <R, T : R> CustomerAdapterResult<T>.getOrElse(
+    onFailure: (cause: Throwable) -> R
+): R {
+    return toResult().getOrElse(onFailure)
+}
+
+@OptIn(ExperimentalCustomerSheetApi::class)
+internal inline fun <R, T> CustomerAdapterResult<T>.map(transform: (value: T) -> R): CustomerAdapterResult<R> {
+    return toResult().map(transform).fromResult()
+}
+
+@OptIn(ExperimentalCustomerSheetApi::class)
+internal inline fun <R, T> CustomerAdapterResult<T>.mapCatching(transform: (value: T) -> R): CustomerAdapterResult<R> {
+    return toResult().mapCatching(transform).fromResult()
+}
+
+@OptIn(ExperimentalCustomerSheetApi::class)
+internal inline fun <T> CustomerAdapterResult<T>.fold(
+    onSuccess: (value: T) -> Unit,
+    onFailure: (cause: Throwable?, displayMessage: String?) -> Unit
+) {
+    return when (this) {
+        is CustomerAdapterResult.Success -> onSuccess(value)
+        is CustomerAdapterResult.Failure -> onFailure(cause, displayMessage)
+    }
+}
+
+internal class CustomerAdapterException(
+    override val cause: Throwable?,
+    val displayMessage: String?,
+) : Exception(displayMessage, cause)
+
+@OptIn(ExperimentalCustomerSheetApi::class)
+internal fun <T> CustomerAdapterResult<T>.toResult(): Result<T> {
+    return when (this) {
+        is CustomerAdapterResult.Success -> Result.success(value)
+        is CustomerAdapterResult.Failure -> Result.failure(
+            CustomerAdapterException(
+                cause = cause,
+                displayMessage = displayMessage
+            )
+        )
+    }
+}
+
+@OptIn(ExperimentalCustomerSheetApi::class)
+internal fun <T> Result<T>.fromResult(): CustomerAdapterResult<T> {
+    return fold(
+        onSuccess = {
+            CustomerAdapterResult.Success(it)
+        },
+        onFailure = {
+            CustomerAdapterResult.Failure(
+                cause = it.cause,
+                displayMessage = it.message
```

Comment:
I don't think we should do this!

Reply:
Why not? It is only coming from `CustomerAdapterException`, but I guess it isn't clear.

Reply:
I have made it private and added an assert